sdk/iot/hub/samples/src/iot_hub_c2d_sample.c

```diff
@@ -0,0 +1,293 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#ifdef _MSC_VER
+// warning C4201: nonstandard extension used: nameless struct/union
+#pragma warning(disable : 4201)
+#endif
+#include <paho-mqtt/MQTTClient.h>
+#ifdef _MSC_VER
+#pragma warning(default : 4201)
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <az_iot_hub_client.h>
+#include <az_result.h>
+#include <az_span.h>
+
+// TODO: #564 - Remove the use of the _az_cfh.h header in samples.
+//              Note: this is required to work-around MQTTClient.h as well as az_span init issues.
+#include <_az_cfg.h>
+
+// Device ID
+#define DEVICE_ID "AZ_IOT_DEVICE_ID"
+
+// IoT Hub FQDN
+#define IOT_HUB_HOSTNAME "AZ_IOT_HUB_HOSTNAME"
+
+// AZ_IOT_DEVICE_X509_CERT_PEM_FILE is the path to a PEM file containing the device certificate and
+// key as well as any intermediate certificates chaining to an uploaded group certificate.
+#define DEVICE_X509_CERT_PEM_FILE "AZ_IOT_DEVICE_X509_CERT_PEM_FILE"
+
+// AZ_IOT_DEVICE_X509_TRUST_PEM_FILE is the path to a PEM file containing the server trusted CA
+// This is usually not needed on Linux or Mac but needs to be set on Windows.
+#define DEVICE_X509_TRUST_PEM_FILE "AZ_IOT_DEVICE_X509_TRUST_PEM_FILE"
+
+#define TIMEOUT_MQTT_DISCONNECT_MS 10 * 1000
+
+static const uint8_t null_terminator = '\0';
+static char device_id[64];
+static char iot_hub_hostname[128];
+static char x509_cert_pem_file[512];
+static char x509_trust_pem_file[256];
+
+static char mqtt_client_id[128];
+static char mqtt_username[128];
+static char mqtt_endpoint[128];
+static az_span mqtt_url_prefix = AZ_SPAN_LITERAL_FROM_STR("ssl://");
+static az_span mqtt_url_suffix = AZ_SPAN_LITERAL_FROM_STR(":8883");
+
+static char c2d_topic[128];
+
+static az_iot_hub_client client;
+static MQTTClient mqtt_client;
+
+static az_result read_configuration_entry(
```

Comment:
General question, which I was even asked by RTOS team, why in our code we do not have comments both in src and these samples.

Reply:
What kind of comments are you looking for? What it will do from a high level? Function details?

Reply:
Comments types :
1. What function does why it will return error.
2. If there are multiple way to do things why we are using this specific approach.
3. Who owns the memory.
4. if function produce some output in string, some possible example of that 

this help us to maintain code.

---
In reply to: [418773973](https://github.com/Azure/azure-sdk-for-c/pull/648#discussion_r418773973) [](ancestors = 418773973)

Reply:
This comment might not be related to this but overall our iot folder

---
In reply to: [418775327](https://github.com/Azure/azure-sdk-for-c/pull/648#discussion_r418775327) [](ancestors = 418775327,418773973)

Reply:
Yup there are definitely improvements to documentation and comments that can be done. And I will be sure to add as many as I can in the coming days. But ultimately this PR is going to sit forever if we wait for things like that. And it'll keep having to be rebased, updated, etc.

I promise I'll continue to document the heck out of it.

Reply:
added to backlog ✔️